pym/bob/input.py

```diff
@@ -2684,17 +2684,20 @@ def generatePackages(self, nameFormatter, envOverrides={}, sandboxEnabled=False)
     def generateTree(self, envOverrides={}, sandboxEnabled=False):
         (env, cacheKey) = self.__getEnvWithCacheKey(envOverrides, sandboxEnabled)
         cacheName = ".bob-tree.dbm"
+        BobLock().getNamedLock(cacheName + ".lock")
 
         # try to load persisted tree
         roots = TreeStorage.load(cacheName, cacheKey)
-        if roots is not None:
-            return roots
 
-        # generate and convert
-        roots = self.__generatePackages(lambda p, m: "unused", env, cacheKey, sandboxEnabled)
+        if roots is None:
+            # generate and convert
+            roots = self.__generatePackages(lambda p, m: "unused", env, cacheKey, sandboxEnabled)
 
-        # save tree cache
-        return TreeStorage.create(cacheName, cacheKey, roots)
+            # save tree cache
+            roots = TreeStorage.create(cacheName, cacheKey, roots)
+
+        BobLock().releaseNamedLock(cacheName + ".lock")
```

Comment:
Hmm, at this stage we still have opened the `TreeStorage` database as read only. If the user changes the recipes and calls Bob in a separate shell then there will still be a read/write conflict. It's not very likely but I think the real solution must use shared/exclusive locks. In this case the lock must be stored in the `TreeStorage` object to persist until the database is closed.